File: shared/src/main/java/ru/privatenull/pnauth/message/AnimatedGradient.java
package ru.privatenull.pnauth.message;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.util.List;

/** Produces one portable animated-gradient frame for every supported message format. */
public final class AnimatedGradient {
    private static final ObjectMapper JSON = new ObjectMapper();

    private AnimatedGradient() {
    }

    public static String frame(String renderedText, MessageFormat format, List<String> colors, int frame, int frameCount) {
        String plain = plain(renderedText, format);
        int[] points = plain.codePoints().toArray();
        if (points.length == 0) return "";
        ArrayNode json = format == MessageFormat.JSON ? JSON.createArrayNode() : null;
        StringBuilder result = new StringBuilder(plain.length() * 16);
        for (int index = 0; index < points.length; index++) {
            String color = color(colors, (index / (double) Math.max(1, points.length - 1)) + frame / (double) frameCount);
            String character = new String(Character.toChars(points[index]));
            switch (format) {
                case MINI_MESSAGE -> result.append('<').append(color).append('>').append(character).append("</").append(color).append('>');
                case LEGACY -> result.append(legacy(color)).append(character);
                case JSON -> {
                    ObjectNode part = json.addObject();
                    part.put("text", character);
                    part.put("color", color);
                }
                case PLAIN -> result.append(character);
            }
        }
        return json == null ? result.toString() : json.toString();
    }

    private static String plain(String value, MessageFormat format) {
        String legacy = MessageRenderers.toLegacy(value == null ? "" : value, format);
        return legacy.replaceAll("(?i)&x(?:&[0-9a-f]){6}", "").replaceAll("(?i)&[0-9a-fk-or]", "");
    }

    private static String color(List<String> colors, double position) {
        double wrapped = position - Math.floor(position);
        double scaled = wrapped * colors.size();
        int first = (int) Math.floor(scaled) % colors.size();
        int second = (first + 1) % colors.size();
        double ratio = scaled - Math.floor(scaled);
        int a = Integer.parseInt(colors.get(first).substring(1), 16);
        int b = Integer.parseInt(colors.get(second).substring(1), 16);
        int red = mix(a >> 16 & 255, b >> 16 & 255, ratio);
        int green = mix(a >> 8 & 255, b >> 8 & 255, ratio);
        int blue = mix(a & 255, b & 255, ratio);
        return String.format("#%02x%02x%02x", red, green, blue);
    }

    private static int mix(int first, int second, double ratio) {
        return (int) Math.round(first + (second - first) * ratio);
    }

    private static String legacy(String color) {
        StringBuilder result = new StringBuilder("&x");
        for (int index = 1; index < color.length(); index++) result.append('&').append(color.charAt(index));
        return result.append("&l").toString();
    }
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/message/AnimatedGradient.kt
package ru.privatenull.pnauth.message

import java.awt.Color
import java.util.Locale

object AnimatedGradient {
    @JvmStatic
    fun frame(
        text: String,
        format: MessageFormat,
        colors: List<String>,
        step: Int,
        width: Int
    ): String {
        if (text.isEmpty() || colors.isEmpty()) return MessageRenderers.forFormat(format).render(text)
        val textLength = text.length
        val numColors = colors.size
        val parsedColors = colors.map { parseColor(it) }

        val builder = StringBuilder()
        if (format == MessageFormat.JSON) {
            builder.append("[")
        }

        val effectiveWidth = width.coerceAtLeast(1)
        for (i in 0 until textLength) {
            val progress = ((i + step) % effectiveWidth).toFloat() / effectiveWidth
            val scaled = progress * (numColors - 1)
            val index = scaled.toInt().coerceIn(0, numColors - 2)
            val factor = scaled - index
            val hexColor = interpolateHex(parsedColors[index], parsedColors[index + 1], factor)
            val char = text[i]

            when (format) {
                MessageFormat.LEGACY -> {
                    builder.append("&x")
                    for (c in hexColor.substring(1)) {
                        builder.append('&').append(c)
                    }
                    builder.append(char)
                }
                MessageFormat.MINI_MESSAGE -> {
                    builder.append("<").append(hexColor).append(">").append(char).append("</").append(hexColor).append(">")
                }
                MessageFormat.JSON -> {
                    if (i > 0) builder.append(",")
                    builder.append("{\"text\":\"").append(char).append("\",\"color\":\"").append(hexColor).append("\"}")
                }
                MessageFormat.PLAIN -> {
                    builder.append(char)
                }
            }
        }
        if (format == MessageFormat.JSON) {
            builder.append("]")
        }
        return builder.toString()
    }

    private fun parseColor(hex: String): Color {
        val c = hex.removePrefix("#")
        val r = c.substring(0, 2).toInt(16)
        val g = c.substring(2, 4).toInt(16)
        val b = c.substring(4, 6).toInt(16)
        return Color(r, g, b)
    }

    private fun interpolateHex(c1: Color, c2: Color, factor: Float): String {
        val r = (c1.red + (c2.red - c1.red) * factor).toInt().coerceIn(0, 255)
        val g = (c1.green + (c2.green - c1.green) * factor).toInt().coerceIn(0, 255)
        val b = (c1.blue + (c2.blue - c1.blue) * factor).toInt().coerceIn(0, 255)
        return String.format(Locale.ROOT, "#%02x%02x%02x", r, g, b)
    }
}
